cilium: force hash recalc before sending packet off to tunnel dev

borkmann opened the pull request:
(See commit desc, this has been observed in Azure envs for off-node traffic hitting k8s service with remote backend where the traffic is then pushed to remote backend via vxlan tunnel.)

borkmann committed:
bpf: Force recalculation of hash for better vxlan/geneve RSS

Reset and recalculate the skb hash at the last point when we set
the tunnel keys, so that when the packet ends up at the vxlan/geneve
device it has a proper L4 hash of the inner packet.

udp_flow_src_port() relies on the skb hash to be always populated
as otherwise we might fall back to a weak hash based on L2 header
or based on some static random id.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gentoo-root commented on bpf/lib/overloadable_skb.h:
```diff
@@ -294,6 +294,16 @@ ctx_set_encap_info(struct __sk_buff *ctx, __u32 src_ip,
+	 * that udp_flow_src_port() actually takes the proper L4 hash
+	 * for source port selection instead of having to fall back.
+	 */
+	if (!get_hash(ctx)) {
```
Checking `skb->hash != 0` seems to be not the best way to check that the hash is missing. It can be present and be zero. Having a 1/65536 chance of slightly increased latency would be a nasty bug.

Instead of these three lines, you can just call `get_hash_recalc(ctx)` straight ahead. It will check `if (!skb->l4_hash && !skb->sw_hash)` internally and calculate the SW hash only if the hash is actually missing.

borkmann replied:
Agree that's better, it would mean that if HW provided an L3 hash then we would still penalize and recalc an L4 hash in SW, but that's probably an acceptable tradeoff. Will change.

borkmann replied:
Thinking about it again, the updated patch should actually not have an effect. udp_flow_src_port() does internally the same skb_get_hash().

borkmann committed:
bpf: Add set_hash_invalid prototype

Add set_hash_invalid prototype so that we can use it out of the
skb-based BPF helper code.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
